js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,387 @@
+                     "Got an unexpected packet without a segment number: " +
+                      data.getName().toUri());
+  }
+  else {
+    try {
+      currentSegment = data.getName().get(-1).toSegment();
+    }
+    catch (ex) {
+      this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+                       "Error decoding the name segment number " +
+                       data.getName().get(-1).toEscapedString() + ": " + ex);
+      return;
+    }
+  }
+
+  // Check for first segment
+  if (this.firstSegmentIsReceived === false) {
+    this.firstSegmentIsReceived = true;
+    if (currentSegment === 0) {
+      this.nextSegmentToRequest += 1;
+    }
+  }
+
+  if (data.getMetaInfo().getFinalBlockId().getValue().size() > 0) {
+    try {
+      this.finalBlockNo = (data.getMetaInfo().getFinalBlockId().toSegment());
+    }
+    catch (ex) {
+      this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+           "Error decoding the FinalBlockId segment number " +
+            data.getMetaInfo().getFinalBlockId().toEscapedString() +
+            ": " + ex);
+      return;
+    }
+  }
+
+  if (this.isDuplicateSegment(currentSegment)) {
+    if (LOG > 3)
+      console.log('Error: duplicate satisfied segment [' + currentSegment + ']');
+    return;
+  }
+  this.satisfiedSegments.push(currentSegment);
+
+  // Save the content
+  this.contentParts[currentSegment] = data.getContent().buf();
+
+  // Check whether we are finished
+  if (this.satisfiedSegments.length >= this.finalBlockNo + 1) {
+    // Concatenate to get content.
+    var content = Buffer.concat(this.contentParts);
+    try {
+      this.onComplete(new Blob(content, false));
+    } catch (ex) {
+      console.log("Error in onComplete: " + NdnCommon.getErrorWithStackTrace(ex));
+    }
+    this.pipelineDelay = (Date.now() - this.pipelineStartTime)/1000;
+
+    if (LOG > 3) {
+      console.log('Pipeline retrieval delay:  ',  this.pipelineDelay);
+      console.log('Total sending Interests delay: ', this.sendInterestDelay);
+      console.log('Total Data verification delay: ', this.onVerifiedDelay);
+      console.log('Total handling Data delay: ', this.onDataDelay);
+    }
+    return;
+  }
+
+  this.decreaseNumberOfSegmentsOnFly();
+  this.onVerifiedDelay += (Date.now() - sTime)/1000;
+
+  // Fetch the next segments
+  this.fetchNextSegments
+    (originalInterest, data.getName());
+};
+
+PipelineFixed.prototype.onValidationFailed = function(data, reason)
+{
+  this.reportError(PipelineFixed.ErrorCode.SEGMENT_VERIFICATION_FAILED,
+                   "Segment verification failed for " + data.getName().toUri() +
+                   " . Reason: " + reason);
+};
+
+PipelineFixed.prototype.onTimeout = function(interest)
+{
+  this.reportError(PipelineFixed.ErrorCode.INTEREST_TIMEOUT,
+                   "Time out for interest " + interest.getName().toUri());
+};
+
+PipelineFixed.prototype.onNack = function(interest)
+{
```

Comment:
It will not happen, because the very first retrieved Data packet should contain the FinalBlockId which tackle both corner problems you've mentioned.

The documentation has been updated to show this requirement (i.e. the first Data needs to contain FinalBlockId)

Reply:
> It will not happen, because the very first retrieved Data packet should contain the FinalBlockId which tackle both corner problems you've mentioned.

False. NDN's FinalBlockId definition is inherited from CCNx 0.8 and never changed. As [clarified by CCNx designers](https://www.lists.cs.ucla.edu/pipermail/ndn-interest/2018-June/002207.html), `boolean IsFinalBlock(content)` is the only universally defined concept. Additionally, [multiple segments could have `IsFinalBlock()` evaluated to true](https://www.lists.cs.ucla.edu/pipermail/ndn-interest/2018-June/002208.html).

> The documentation has been updated to show this requirement (i.e. the first Data needs to contain FinalBlockId)

You can't arbitrary add requirements beyond what network layer protocol guarantees. If you want to change network layer protocol, the protocol change needs to be committed to NDN-TLV repository first before any code could rely on that.

Reply:
Ok, this is gonna be a problem, then...
So how we can distinguish a Nack triggered by a problem and a Nack triggered by window-size>finalBlockId??

Btw, JS using timers and scheduler is not highly recommended as from one browser to another it differs (https://medium.freecodecamp.org/javascript-timers-everything-you-need-to-know-5f31eaa37162). Right now, we do not schedule the Interests, we just issue them and pipeline the next ones as we receive Data packet for previous Interests. Is there any best practice in JS for such scenarios, though??

Reply:
> So how we can distinguish a Nack triggered by a problem and a Nack triggered by window-size>finalBlockId??

As soon as you receive a segment where `IsFinalBlock()` evaluates to true, invoke `Face.prototype.removePendingInterest` to cancel all segment Interests beyond the final segment. Afterwards, you won't receive any more Nacks.
If you get Nacks before receiving a segment where `IsFinalBlock()` evaluates to true, unfortunately they are not distinguishable.

Reply:
> If you get Nacks before receiving a segment where IsFinalBlock() evaluates to true, unfortunately they are not distinguishable.

This is the case I was thinking about... So, there is not way to handle this??